File: 00_The_Programming_Contest_Training_Manual/1.6.1_Collatz_Conjecture/main.py
#!/usr/bin/python3


def hailstones(start_n: int) -> int:
    """
    Execute one step for the collatz conjecture.
    """
    while True:
        if start_n == 1:
            break

        elif start_n % 2 == 0:
            start_n //= 2

        else:
            start_n = 3 * start_n + 1

        yield start_n


def max_hail_cycle(start_h: int, end_h: int) -> int:
    """
    Find the maximum cycle length of all the starting numbers in the envelope
    defined by 'start_h' and 'end_h'.
    """

    # Record all the cycles seen by the function
    max_cycles = {}
    max_c = 0

    # Test each possible hailstone starting number
    for hail in range(start_h, end_h + 1):
        temp_max = 1

        # Go through each individual hail
        for i in hailstones(hail):
            # Check for cycles that have been seen before
            if i in max_cycles:
                temp_max += max_cycles[i]
                break
            else:
                temp_max += 1

        # Save the result
        max_cycles[hail] = temp_max

        # Set the new maximum
        if temp_max > max_c:
            max_c = temp_max

    return max_c


if __name__ == "__main__":
    assert max_hail_cycle(1, 10) == 20
    assert max_hail_cycle(100, 200) == 125
    assert max_hail_cycle(201, 210) == 89
    assert max_hail_cycle(900, 1000) == 174

File: 00_The_Programming_Contest_Training_Manual/1.6.1_Collatz_Conjecture/src/main.rs
use std::collections::HashMap;

/// Execute one step in the hailstone cycle.
fn hail_step(start_n: u32) -> u32 {
    if start_n % 2 == 0 {
        return start_n / 2;
    } else {
        return 3 * start_n + 1;
    }
}

/// Find the maximum hailstone cycles for a range of numbers.
fn max_hail_cycles(start_h: u32, end_h: u32) -> u32 {
    let mut max_cycle = 0;
    let mut prev_cycles: HashMap<u32, u32> = HashMap::new();

    for initial_h in start_h..=end_h {
        let mut tmp_cycle_cnt = 1;
        let mut tmp_hail = initial_h;

        /* Run the whole hail stone path. */
        while tmp_hail != 1 {
            tmp_hail = hail_step(tmp_hail);

            /* Check to see if the hail path has seen before */
            if prev_cycles.contains_key(&tmp_hail) {
                tmp_cycle_cnt += prev_cycles[&tmp_hail];
                break;
            } else {
                tmp_cycle_cnt += 1;
            }
        }

        /* Test for a new highest cycle. */
        if tmp_cycle_cnt > max_cycle {
            max_cycle = tmp_cycle_cnt;
        }

        /* Save the cycle length of the current start number. */
        prev_cycles.insert(initial_h, tmp_cycle_cnt);
    }

    return max_cycle;
}

fn main() {
    assert_eq!(max_hail_cycles(1, 10), 20);
    assert_eq!(max_hail_cycles(100, 200), 125);
    assert_eq!(max_hail_cycles(201, 210), 89);
    assert_eq!(max_hail_cycles(900, 1000), 174);
}
